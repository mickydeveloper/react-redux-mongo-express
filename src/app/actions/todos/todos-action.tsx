import axios, { AxiosPromise, AxiosResponse, AxiosError } from "axios";
import { Dispatch } from "redux";
import { ThunkAction } from "redux-thunk";
import { toast } from "react-toastify";
import { API_SETTINGS } from "../../utils/api/settings";
import { ACTION_TYPE } from "../../actions/actions";
import { assignHeader } from "../../utils/api/headers";
//INTERFACES
import { ITodo } from "./todos.d";
import { IReducers } from "./../../reducers/index.d";
//CONST
import { TOASTER } from "./../../utils/constants/toasters";

const apiurl: string = `${API_SETTINGS.API_URL}/todos`;

export function getTodos(): ThunkAction<
  Promise<ITodo | string>,
  IReducers,
  null
> {
  return function(dispatch: Dispatch<IReducers>): Promise<ITodo | string> {
    return axios
      .get(`${apiurl}`, assignHeader())
      .then((response: AxiosResponse): Promise<ITodo> => {
        dispatch({
          type: ACTION_TYPE.TODO.GET,
          todos: response.data
        });
        return response.data;
      })
      .catch((error: AxiosError): Promise<string> => {
        toast.error(TOASTER.API.ERROR.GET);
        return error.response
          ? Promise.reject(error.response.data.message)
          : Promise.reject("Could not retrieve todos");
      });
  };
}

export function createTodo(
  todo: ITodo
): ThunkAction<Promise<ITodo | string>, IReducers, null> {
  return function(dispatch: Dispatch<IReducers>): Promise<ITodo | string> {
    return axios
      .post(`${apiurl}`, todo, assignHeader())
      .then((response: AxiosResponse): Promise<ITodo> => {
        /* Right now there is not need to dispatch this action as on the todo list view we make get request on each view load
        dispatch({
          type: ACTION_TYPE.TODO.CREATE,
          todo
        });*/
        toast.success(TOASTER.API.SUCCESS.CREATED);
        return response.data;
      })
      .catch((error: AxiosError): Promise<string> => {
        toast.error(TOASTER.API.ERROR.POST);
        return error.response
          ? Promise.reject(error.response.data.message)
          : Promise.reject("Could not create todo");
      });
  };
}

export function updateTodo(
  todoId: string,
  text: string
): ThunkAction<Promise<ITodo | string>, IReducers, null> {
  return function(dispatch: Dispatch<IReducers>): Promise<ITodo | string> {
    return axios
      .put(`${apiurl}/` + todoId, { text }, assignHeader())
      .then((response: AxiosResponse): Promise<ITodo> => {
       /* Right now there is not need to dispatch this action as on the todo list view we make get request on each view load
        dispatch({
          type: ACTION_TYPE.TODO.UPDATE,
          todoId,
          text
        });
        */
       toast.info(TOASTER.API.SUCCESS.UPDATED);
        return response.data;
      })
      .catch((error: AxiosError): Promise<string> => {
        toast.error(TOASTER.API.ERROR.PUT);
        return error.response
          ? Promise.reject(error.response.data.message)
          : Promise.reject("Could not update todo");
      });
  };
}

export function deleteTodo(
  id: string
): ThunkAction<Promise<ITodo | string>, IReducers, null> {
  return function(dispatch: Dispatch<IReducers>): Promise<ITodo | string> {
    return axios
      .delete(`${apiurl}/` + id, assignHeader())
      .then((response: AxiosResponse): Promise<ITodo> => {
        dispatch({
          type: ACTION_TYPE.TODO.DELETE,
          id
        });
        toast.warn(TOASTER.API.SUCCESS.DELETED);
        return response.data;
      })
      .catch((error: AxiosError): Promise<string> => {
        toast.error(TOASTER.API.ERROR.DELETE);
        return error.response
          ? Promise.reject(error.response.data.message)
          : Promise.reject("Could not delete todo");
      });
  };
}

export function setCurrentTodo(id: string, text: string): Function {
  return function(dispatch: Dispatch<IReducers>): Promise<ITodo> {
    return new Promise((resolve, reject) =>
      dispatch({
        type: ACTION_TYPE.TODO.CURRENT,
        todo: {
          id,
          text
        }
      })
    );
  };
}
