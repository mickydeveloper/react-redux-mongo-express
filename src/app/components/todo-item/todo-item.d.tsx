export interface ITodoProps {
    id: string;
    text: string;
    deleteTodo: Function;
    setCurrentTodo: Function;
}