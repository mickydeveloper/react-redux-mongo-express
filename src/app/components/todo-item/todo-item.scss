.todo-item {
  display: grid;
  grid-template-columns: auto 20px;
  margin-bottom: 0.8rem;
  background: #f2f2f2;
  align-items: center;
  padding: 0.8rem;
  border-radius: 5px;

  h5 {
    margin: 0;
  }

  i {
    line-height: 1rem;
    font-size: 1.4rem;
    color: $color-valencia;
    transition: all 0.2s ease-in-out;
  }
}
