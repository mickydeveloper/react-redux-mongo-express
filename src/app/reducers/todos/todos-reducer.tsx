import { ACTION_TYPE } from "../../actions/actions";
import { Reducer } from "redux";
import { ITodo } from "../../actions/todos/todos.d";

export const todosReducer: Reducer<ITodo[] | undefined> = function(
  state: ITodo[] = [],
  action: any
): ITodo[] | undefined {
  switch (action.type) {
    case ACTION_TYPE.TODO.GET:
      return [...action.todos];
    case ACTION_TYPE.TODO.DELETE:
      return state.filter(todo => todo.id !== action.id);
  }
  return state;
};
