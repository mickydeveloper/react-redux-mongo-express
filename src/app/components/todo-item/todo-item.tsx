import React from "react";
import { Link } from "react-router-dom";
import { ITodoProps } from "./todo-item.d";

export const TodoItem: React.StatelessComponent<ITodoProps> = (
  props: ITodoProps
) => {
  return (
    <div className="todo-item">
      <Link to={"/update/" + props.id} onClick={() => props.setCurrentTodo(props.id, props.text)}><h5>{props.text}</h5></Link>
      <i className="fa fa-trash-o" onClick={() => props.deleteTodo(props.id)}/>
    </div>
  );
};
