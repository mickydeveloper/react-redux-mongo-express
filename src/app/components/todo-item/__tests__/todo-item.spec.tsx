import * as React from "react";
import { shallow, mount } from "enzyme";
import { TodoItem } from "../todo-item";
import { deleteTodo, setCurrentTodo } from "../../../actions/todos/todos-action";

describe("testing <TodoItem>", () => {
    it("the snapshot should match", () => {
        const todo = {
            id: "868et8gdf8g8sdg",
            text: "This is just a test"
        };
        const result = shallow(<TodoItem id={todo.id} text={todo.text} deleteTodo={deleteTodo} setCurrentTodo={setCurrentTodo}/>);
        expect(result).toMatchSnapshot();
    });
});