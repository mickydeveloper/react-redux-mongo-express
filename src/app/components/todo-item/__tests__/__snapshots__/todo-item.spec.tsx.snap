// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing <TodoItem> the snapshot should match 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    deleteTodo={[Function]}
    id="868et8gdf8g8sdg"
    setCurrentTodo={[Function]}
    text="This is just a test"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Link
          onClick={[Function]}
          replace={false}
          to="/update/868et8gdf8g8sdg"
        >
          <h5>
            This is just a test
          </h5>
        </Link>,
        <i
          className="fa fa-trash-o"
          onClick={[Function]}
        />,
      ],
      "className": "todo-item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <h5>
            This is just a test
          </h5>,
          "onClick": [Function],
          "replace": false,
          "to": "/update/868et8gdf8g8sdg",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "This is just a test",
          },
          "ref": null,
          "rendered": "This is just a test",
          "type": "h5",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "fa fa-trash-o",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "i",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Link
            onClick={[Function]}
            replace={false}
            to="/update/868et8gdf8g8sdg"
          >
            <h5>
              This is just a test
            </h5>
          </Link>,
          <i
            className="fa fa-trash-o"
            onClick={[Function]}
          />,
        ],
        "className": "todo-item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <h5>
              This is just a test
            </h5>,
            "onClick": [Function],
            "replace": false,
            "to": "/update/868et8gdf8g8sdg",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "This is just a test",
            },
            "ref": null,
            "rendered": "This is just a test",
            "type": "h5",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "fa fa-trash-o",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "i",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
